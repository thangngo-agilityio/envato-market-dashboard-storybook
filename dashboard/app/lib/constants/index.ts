export * from './routers';
export * from './time';
export * from './images';
export * from './form';
export * from './messages';
export * from './regex';
export * from './texts';
export * from './themes';
export * from './api';
export * from './images';
export * from './menu';
export * from './sidebar';
export * from './charts';
export * from './sidebar';
export * from './common';
export * from './options';
export * from './transaction';
export * from './store';
export * from './status';
export * from './pagination';
export * from './columns';
export * from './faq';
export * from './avatar';
export * from './sizes';
export * from './setting';
export * from './termAndCondition';
export * from './notification';
export * from './authentication';
export * from './discount';
export * from './metadata';
export * from './og-properties';
export * from './chat';
export * from './product';
