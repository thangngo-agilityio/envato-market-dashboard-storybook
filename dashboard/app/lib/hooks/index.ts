export * from './useForm';
export * from './useAuth';
export * from './useStatistical';
export * from './useTransactions';
export * from './useDebounce';
export * from './useSearch';
export * from './usePagination';
export * from './useStore';
export * from './useUser';
export * from './usePinCode';
export * from './useNotification';
export * from './useWallet';
export * from './useMoney';
export * from './useChat';
export * from './useImageUploader';
export * from './useProducts';
export * from './useRecentActivities';
