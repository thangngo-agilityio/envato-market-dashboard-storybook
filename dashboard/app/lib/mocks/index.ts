export * from './spending';
export * from './table';
export * from './transaction';
export * from './product';
export * from './user';
export * from './time';
export * from './statistic';
export * from './chat';
export * from './issues';
export * from './wallet';
export * from './supports';
export * from './recentActivities';
