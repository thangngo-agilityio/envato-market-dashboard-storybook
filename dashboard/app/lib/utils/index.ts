export * from './transaction';
export * from './product';
export * from './time';
export * from './sort';
export * from './user';
export * from './url';
export * from './helpers';
export * from './text';
export * from './localStorage';
export * from './image';
export * from './prefetch';
export * from './toast';
export * from './getErrorMessage';
export * from './firebase';
export * from './recentActivities';
export * from './logAcitvity';
