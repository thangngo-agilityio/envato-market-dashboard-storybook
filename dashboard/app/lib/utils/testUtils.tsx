import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

const queryClient = new QueryClient();

export const renderQueryProviderTest = (component: JSX.Element) => {
  const { render } = testLibReactUtils;
  return render(
    <QueryClientProvider client={queryClient}>{component}</QueryClientProvider>,
  );
};
