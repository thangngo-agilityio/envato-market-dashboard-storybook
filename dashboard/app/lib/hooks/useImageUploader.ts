import { useCallback } from 'react';

// Constants
import { ERROR_MESSAGES, LIMIT_PRODUCT_IMAGES, REGEX } from '@/lib/constants';

// Services
import { uploadImage } from '@/lib/services';

interface ImageUploaderProps {
  onChange: (imagesUpload: string[]) => void;
  onUploadError: (errorMessage: string) => void;
  setPreviewURL: React.Dispatch<React.SetStateAction<string[]>>;
}

export const useImageUploader = ({
  onChange,
  onUploadError,
  setPreviewURL,
}: ImageUploaderProps) => {
  const onDrop = useCallback(
    (acceptedFiles: File[]) => {
      const imagesPreview: React.SetStateAction<string[]> = [];
      const imagesUpload: string[] = [];

      if (acceptedFiles.length > LIMIT_PRODUCT_IMAGES) {
        return onUploadError(ERROR_MESSAGES.UPLOAD_IMAGE_ITEM);
      }

      acceptedFiles.map(async (file) => {
        if (!file) {
          return;
        }

        // Check type of image
        if (!REGEX.IMG.test(file.name)) {
          return onUploadError(ERROR_MESSAGES.UPLOAD_IMAGE);
        }

        // Uploading file
        try {
          onChange([]);
          const previewImage: string = URL.createObjectURL(file);
          const formData = new FormData();

          formData.append('image', file);
          imagesPreview.push(previewImage);

          const result = await uploadImage(formData);
          imagesUpload.push(result);
        } catch (error) {
          onUploadError(ERROR_MESSAGES.UPDATE_FAIL.title);
        }
        setPreviewURL(imagesPreview);
      });

      onChange(imagesUpload);
    },
    [onChange, onUploadError, setPreviewURL],
  );

  return { onDrop };
};

export default useImageUploader;
