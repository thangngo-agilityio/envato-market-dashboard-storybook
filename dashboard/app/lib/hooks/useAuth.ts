import { useRouter } from 'next/navigation';
import { useCallback, useState } from 'react';
import { AxiosError, AxiosResponse } from 'axios';
import { shallow } from 'zustand/shallow';
import { getAuth, signInWithEmailAndPassword } from 'firebase/auth';

// Constants
import {
  END_POINTS,
  ERROR_MESSAGES,
  IMAGES,
  LOGOUT_TIME,
  ROUTES,
} from '@/lib/constants';

// Services
import { MainHttpService, recentActivitiesHttpService } from '@/lib/services';

// Types
import { TUserDetail, EActivity, TActivitiesRequest } from '@/lib/interfaces';

// Utils
import { formatUppercaseFirstLetter, getCurrentTimeSeconds } from '@/lib/utils';

// Stores
import { authStore } from '@/lib/stores';

type TSignUpErrorField = Partial<
  Record<keyof Omit<TUserDetail, 'id' | 'createdAt'>, string>
>;

export type TUserInfo = Omit<TUserDetail, 'password'> | null;

export type TUserAxiosResponse = Omit<
  TUserDetail & { _id: string },
  'id'
> | null;

export type TUseAuth = {
  user: TUserInfo;
  isRemember: boolean;
  date: number;
  setUser: (user: TUserDetail) => void;
  signIn: (
    {
      email,
      password,
    }: {
      email: string;
      password: string;
    },
    isRemember?: boolean,
  ) => Promise<void>;
  signUp: (userInfo: Omit<TUserDetail, 'id' | 'createdAt'>) => Promise<{
    errors?: TSignUpErrorField;
  }>;
  signOut: () => void;
};

export const useAuth = () => {
  const { user } = authStore();
  const [isLogout, setIsLogout] = useState(false);
  const router = useRouter();
  const { updateStore, clearStore } = authStore(
    (state) => ({
      updateStore: state.updateStore,
      clearStore: state.clearStore,
    }),
    shallow,
  );
  const handleSignInWithFirebase = useCallback(
    async (email: string, password: string) => {
      const auth = getAuth();
      return await signInWithEmailAndPassword(auth, email, password).then(
        (userCredential) => {
          const { uid, email } = userCredential.user;

          return {
            uid,
            email,
          };
        },
      );
    },
    [],
  );

  const handleSignIn = useCallback(
    async (
      {
        email,
        password,
        fcmToken = '',
      }: {
        email: string;
        password: string;
        fcmToken?: string;
      },
      isRemember?: boolean,
    ): Promise<void> => {
      try {
        const { uid } = await handleSignInWithFirebase(email, password);

        const { data }: AxiosResponse<TUserAxiosResponse | undefined> =
          await MainHttpService.post<TUserAxiosResponse | undefined>(
            `${END_POINTS.SIGN_IN}`,
            {
              email,
              password,
              fcmToken,
              uid,
            },
            {},
            EActivity.SIGN_IN,
          );

        let localData: TUserInfo | undefined;
        if (data) {
          const { _id, ...rest } = data;
          localData = { ...rest, id: _id };
        }
        return updateStore({
          user: localData,
          isRemember,
          date: getCurrentTimeSeconds(),
        });
      } catch (error) {
        const { response } = error as AxiosError<string>;

        if (!response) {
          throw new Error(
            formatUppercaseFirstLetter(ERROR_MESSAGES.AUTH_INCORRECT),
          );
        }

        throw new Error(formatUppercaseFirstLetter(response?.data));
      }
    },
    [handleSignInWithFirebase, updateStore, user?.id],
  );

  const handleSignUp = useCallback(
    async (
      userInfo: Omit<TUserDetail, 'id' | 'createdAt'> & { fcmToken: string },
    ): Promise<void> => {
      const { email, password, firstName, lastName, fcmToken } = userInfo;
      try {
        const { data }: AxiosResponse<TUserAxiosResponse | undefined> =
          await MainHttpService.post<TUserAxiosResponse | undefined>(
            `${END_POINTS.SIGN_UP}`,
            {
              ...userInfo,
              avatarURL: IMAGES.AVATAR_SIGN_UP.url,
              createdAt: Date.now(),
              email,
              password,
              firstName,
              lastName,
              fcmToken,
            },
            {},
            EActivity.SIGN_UP,
          );

        let localData: TUserInfo | undefined;
        if (data) {
          const { _id, ...rest } = data;
          localData = { ...rest, id: _id };
        }

        return updateStore({ user: localData, date: getCurrentTimeSeconds() });
      } catch (error) {
        const { response } = error as AxiosError<string>;

        throw new Error(formatUppercaseFirstLetter(response?.data));
      }
    },
    [updateStore, user?.id],
  );

  const handleLogout = useCallback(
    async (
      redirectPath?: string,
      option?: keyof Pick<typeof router, 'push' | 'replace'>,
    ) => {
      setIsLogout(true);

      await recentActivitiesHttpService.post<TActivitiesRequest>(
        END_POINTS.RECENT_ACTIVITIES,
        {
          userId: user?.id,
          actionName: EActivity.SIGN_OUT,
        },
      );

      setTimeout(() => {
        clearStore();
        setIsLogout(false);
        router[option ?? 'replace'](redirectPath ?? ROUTES.LOGIN);
      }, LOGOUT_TIME);
    },
    [clearStore, router, user?.id],
  );

  return {
    isLogoutHandling: isLogout,
    setUser: updateStore,
    signIn: handleSignIn,
    signUp: handleSignUp,
    signOut: handleLogout,
    handleSignInWithFirebase,
  };
};
