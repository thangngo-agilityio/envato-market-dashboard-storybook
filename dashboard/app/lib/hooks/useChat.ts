import { Dispatch, MutableRefObject, SetStateAction, useEffect } from 'react';

// Stores
import { authStore } from '@/lib/stores';

// Interface
import { TUserInfo, useGetUserDetails } from '.';
import { TMessages } from '@/lib/interfaces';

// Firebase
import { collection, doc, getDocs, onSnapshot } from 'firebase/firestore';

// Constants
import { AUTHENTICATION_ROLE, FIREBASE_CHAT } from '@/lib/constants';

// Utils
import { db } from '@/lib/utils';
import { MOCK_SUPER_ADMIN } from '../mocks';
import { getAdminDetailsWithId } from '../services';

// TODO: if have real id from firestore
export const useGetRoomChat = () => {
  const user = authStore((state) => state.user);
  const { filterDataUser } = useGetUserDetails(user?.id || '');

  const isAdmin = filterDataUser?.find(
    (user) => user.role === AUTHENTICATION_ROLE.SUPER_ADMIN,
  );

  const adminUserId = isAdmin?._id;
  const comBindIdChat = `${user?.id || ''}${adminUserId}`;
  return comBindIdChat;
};

// Author: Loc Vo
export const getInfoRoomChat = async (user: TUserInfo) => {
  const admin = await getAdminDetailsWithId(user?.id);
  const userUid = user?.uid;

  return {
    roomChatId: `${admin.uid}${userUid}`,
    userId: userUid || '',
    adminId: admin.uid || '',
    avatarUrl: user?.avatarURL || '',
    avatarAdminUrl: MOCK_SUPER_ADMIN?.avatarURL || '',
    displayName: `${user?.firstName} ${user?.lastName} `,
  };
};

export const getLists = async () => {
  const usersCollection = collection(db, FIREBASE_CHAT.USER_CHATS);
  const usersSnapshot = await getDocs(usersCollection);

  const chatList = usersSnapshot.docs
    .map((doc) => ({
      id: doc.id,
      data: doc.data(),
    }))
    .map((item) => Object.values(item.data)[0]) as [];

  return {
    chatList,
  };
};

export const useSubscribeToChat = (
  roomId: string,
  setMessages: Dispatch<SetStateAction<TMessages[]>>,
  boxRef?: MutableRefObject<HTMLDivElement | null>,
) => {
  useEffect(() => {
    if (!roomId) return;

    const unSub = onSnapshot(
      doc(db, FIREBASE_CHAT.CHATS, roomId),
      async (doc) => {
        doc.exists() && setMessages(doc.data().messages);
      },
    );

    if (boxRef?.current) {
      boxRef.current.scrollTop = boxRef.current.scrollHeight;
    }

    return () => {
      unSub();
    };
  }, [setMessages, roomId, boxRef]);
};
