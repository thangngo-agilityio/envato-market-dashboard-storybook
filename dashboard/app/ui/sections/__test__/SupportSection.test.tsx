import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Utils
import { renderQueryProviderTest } from '@/lib/utils/testUtils';

// Sections
import { SupportsSection } from '..';

// Hooks
import { useCreateIssues } from '@/lib/hooks';

// Mocks
import { ISSUES_LIST } from '@/lib/mocks';

jest.mock('@/lib/hooks', () => ({
  ...jest.requireActual('@/lib/hooks'),
  useCreateIssues: jest.fn(),
  useGetListIssues: jest.fn(() => ({
    data: [],
    isFetchingNextPage: false,
    hasNextPage: false,
    fetchNextPage: jest.fn(),
    isFetching: false,
  })),
}));

jest.mock('@chakra-ui/react', () => ({
  ...jest.requireActual('@chakra-ui/react'),
  useToast: () => jest.fn(),
}));

const createIssuesMock = jest.fn();

describe('SupportsSection render', () => {
  beforeAll(() => {
    (useCreateIssues as jest.Mock).mockReturnValue({
      data: ISSUES_LIST,
      useCreateIssues: createIssuesMock,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Should render match with snapshot.', () => {
    const { container } = renderQueryProviderTest(<SupportsSection />);

    expect(container).toMatchSnapshot();
  });

  it('calls createIssues with correct parameters and displays success toast on success', async () => {
    const createIssuesMock = jest
      .fn()
      .mockImplementation((issueData, { onSuccess }) => {
        onSuccess();
      });

    (useCreateIssues as jest.Mock).mockReturnValue({
      mutate: createIssuesMock,
      status: 'idle',
      isPending: false,
    });

    render(<SupportsSection />);

    // Simulate user input
    await userEvent.type(screen.getByLabelText('First Name'), 'John');
    await userEvent.type(screen.getByLabelText('Last Name'), 'Doe');
    await userEvent.type(
      screen.getByLabelText('Email'),
      'john.doe@example.com',
    );
    await userEvent.type(screen.getByLabelText('Title'), 'Test Issue');
    await userEvent.type(
      screen.getByText(/description/i),
      'This is a test description for an issue.',
    );

    await userEvent.click(screen.getByTestId('btn-create-issues'));

    expect(createIssuesMock).toHaveBeenCalled();
  });

  it('displays error toast when createIssues fails', async () => {
    const createIssuesMock = jest
      .fn()
      .mockImplementation((issueData, { onError }) => {
        onError();
      });

    (useCreateIssues as jest.Mock).mockReturnValue({
      mutate: createIssuesMock,
      status: 'idle',
      isPending: false,
    });

    const useToastMock = jest.fn();
    useToastMock.mockReturnValue(jest.fn());

    render(<SupportsSection />);

    await userEvent.type(screen.getByLabelText('First Name'), 'Jane');
    await userEvent.type(screen.getByLabelText('Last Name'), 'Doe');
    await userEvent.type(
      screen.getByLabelText('Email'),
      'jane.doe@example.com',
    );
    await userEvent.type(screen.getByLabelText('Title'), 'Test Issue Failure');
    await userEvent.type(
      screen.getByText(/description/i),
      'This should trigger an error.',
    );

    await userEvent.click(screen.getByTestId('btn-create-issues'));

    expect(createIssuesMock).toHaveBeenCalled();
  });
});
