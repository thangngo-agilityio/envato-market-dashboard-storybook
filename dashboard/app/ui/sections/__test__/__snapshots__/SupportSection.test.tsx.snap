// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SupportsSection render Should render match with snapshot. 1`] = `
<div>
  <div
    class="css-ebwd27"
  >
    <div
      class="css-0"
    >
      <form
        class="chakra-stack css-nx6amm"
        id="register-form"
      >
        <div
          class="css-at8f8g"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1rh809e"
              for="field-:r0:"
              id="field-:r0:-label"
            >
              First Name
            </label>
            <div
              class="chakra-input__group css-1091msq"
              data-group="true"
            >
              <input
                class="chakra-input css-1js16q1"
                id="field-:r0:"
                name="firstName"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1rh809e"
              for="field-:r1:"
              id="field-:r1:-label"
            >
              Last Name
            </label>
            <div
              class="chakra-input__group css-1091msq"
              data-group="true"
            >
              <input
                class="chakra-input css-vyd2b0"
                id="field-:r1:"
                name="lastName"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="css-6w5lc"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1rh809e"
              for="field-:r2:"
              id="field-:r2:-label"
            >
              Email
            </label>
            <div
              class="chakra-input__group css-1091msq"
              data-group="true"
            >
              <input
                class="chakra-input css-1js16q1"
                id="field-:r2:"
                name="email"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1rh809e"
              for="field-:r3:"
              id="field-:r3:-label"
            >
              Phone Number (optional)
            </label>
            <div
              class="chakra-input__group css-1091msq"
              data-group="true"
            >
              <input
                class="chakra-input css-vyd2b0"
                id="field-:r3:"
                name="phoneNumber"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <p
          class="chakra-text css-18nurfg"
        >
          Support issues
        </p>
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-1rh809e"
            for="field-:r4:"
            id="field-:r4:-label"
          >
            Title
          </label>
          <div
            class="chakra-input__group css-1091msq"
            data-group="true"
          >
            <input
              class="chakra-input css-vyd2b0"
              id="field-:r4:"
              name="title"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="css-ipp4t7"
        >
          <p
            class="chakra-text css-16g65kx"
          >
            Description
          </p>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <div
              class="css-1yae3jf"
            />
          </div>
        </div>
        <button
          aria-label="btn submit-ticket"
          class="chakra-button css-18sb6wm"
          data-testid="btn-create-issues"
          disabled=""
          type="submit"
        >
          Submit Ticket
        </button>
      </form>
    </div>
    <div
      class="css-n8vry8"
    >
      <p
        class="chakra-text css-9y7ik"
      >
        Recent Support
      </p>
      <div
        class="css-1t02a7v"
      />
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
